File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/youtube/YoutubeUpdateProcessor.java
package com.github.bumblebee.command.youtube;

import com.github.bumblebee.command.youtube.entity.AtomFeed;
import com.github.bumblebee.command.youtube.entity.PostedVideo;
import com.github.bumblebee.command.youtube.service.YoutubeSubscriptionService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import com.github.telegram.api.BotApi;

import java.util.*;

@Component
public class YoutubeUpdateProcessor {

    private static final String VIDEO_URL = "https://www.youtube.com/watch?v=";

    private final YoutubeSubscriptionService service;
    private final BotApi botApi;
    private final List<PostedVideo> postedVideos;

    @Autowired
    public YoutubeUpdateProcessor(BotApi botApi, YoutubeSubscriptionService service) {
        this.botApi = botApi;
        this.service = service;
        this.postedVideos = service.retrievePostedVideos();
    }

    public void process(AtomFeed feed) {
        String videoId = feed.getEntry().getVideoId();
        if (!postedVideos.stream().anyMatch(postedVideos -> postedVideos.getVideoId().equals(videoId))) {
            postVideo(feed);
            markVideoAsPosted(videoId);
        }
    }

    private void postVideo(AtomFeed feed) {
        for (Long chatId : service.getChatIds(feed.getEntry().getChannelId())) {
            botApi.sendMessage(chatId, VIDEO_URL + feed.getEntry().getVideoId());
        }
    }

    private void markVideoAsPosted (String videoId) {
        PostedVideo postedVideo = new PostedVideo();
        postedVideo.setVideoId(videoId);
        postedVideo.setPostedDate(new Date());
        postedVideos.add(postedVideo);
        service.addPostedVideo(postedVideo);
    }

}

File: telegram-bot-bumblebee-core/src/main/kotlin/com/github/bumblebee/command/youtube/YoutubeUpdateProcessor.kt
package com.github.bumblebee.command.youtube

import com.github.bumblebee.command.youtube.entity.PostedVideo
import com.github.bumblebee.command.youtube.entity.VideoNotification
import com.github.bumblebee.command.youtube.service.YoutubeSubscriptionService
import com.github.bumblebee.util.logger
import com.github.telegram.api.BotApi
import org.springframework.stereotype.Component
import java.util.*

@Component
class YoutubeUpdateProcessor(private val botApi: BotApi,
                             private val service: YoutubeSubscriptionService) {
    private val postedVideos: MutableList<PostedVideo> = service.retrievePostedVideos().toMutableList()

    fun process(videoNotification: VideoNotification) {
        if (!postedVideos.any { it.videoId == videoNotification.videoId }) {
            postVideo(videoNotification)
            markVideoAsPosted(videoNotification.videoId)
        }
    }

    private fun postVideo(videoNotification: VideoNotification) {
        log.info("Posting video: {}", videoNotification)
        service.getChatIds(videoNotification.channelId).forEach { chatId ->
            val url = VIDEO_URL + videoNotification.videoId
            log.info("Posting video to chat {}: {}", chatId, url)
            botApi.sendMessage(chatId, url)
        }
    }

    private fun markVideoAsPosted(videoId: String) {
        val video = PostedVideo(videoId, Date())
        postedVideos.add(video)
        service.addPostedVideo(video)
    }

    companion object {
        val log = logger<YoutubeUpdateProcessor>()
        private val VIDEO_URL = "https://www.youtube.com/watch?v="
    }
}
